sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorConnection.java

```diff
@@ -441,15 +442,8 @@ public void dispose() {
                 .log("Unable to emit shutdown signal.");
         }
 
-        final Mono<Void> cbsCloseOperation;
-        if (cbsChannelProcessor != null) {
-            cbsCloseOperation = cbsChannelProcessor.flatMap(channel -> channel.closeAsync());
-        } else {
-            cbsCloseOperation = Mono.empty();
-        }
-
-        final Mono<Void> managementNodeCloseOperations = Mono.when(
-            Flux.fromStream(managementNodes.values().stream()).flatMap(node -> node.closeAsync()));
+        // Note that we don't need to manually close cbs and management nodes here as RequestResponseChannel
+        // has subscriber to close itself after shutdown signal emitted.
```

Comment:
Another two reasons for remove duplicate closing here are:
1.  Timeout Exception will be thrown. Because when closing `RequestResponseChannel`, it has timeout to wait send and receive links are completed close, but closing channel here will block the `closeConnectionWork` task that could execute `onLinkFinal` method to complete links. It cause the scenario like deadlock and in the end got timeout exception. 
1.   After we added `takeUtilOther(shutdownSignal)`, the `cbsChannelProcessor` and `managementNode` can't subscribe with the initial `RequestResponseChannel` after shutdown signal emitted. An IllegalStateException will thrown if we try to close `cbsChannelProcessor` here.

Reply:
thanks, Kun.

Regarding 1 - my understanding of the code flow is - 


1. The RRChannel first "initiates" the closure of [links](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RequestResponseChannel.java#L261-L262) which sends Detach frame to service.
2. Service responds with Detach frame acknowledging the 1, which is "happy-path finalization".
3. Because we may never get a Detach response from the service, the RRChannel [associates](https://github.com/Azure/azure-sdk-for-java/blob/a2c852f6c8f6eed6034342f008093d9fb6bddc68/sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RequestResponseChannel.java#L270) "timeout-path finalization" to step2.
4. Either of the "finalization" (happy-path, timeout-path) is guaranteed to be executed so that the "cleanup" phase runs.
5. The "cleanup" phase is where we complete the pending Monos with error. Which is important for callers who are asynchronously blocking on sendWithAck to resume; otherwise it can result in unrecovered receivers.

The step-2 happens in response to Reactor-thread reacting to the "Read-event raised by the NW SocketChannel". 

IIRC that's the closure flow of RRChannel type (if we treat it as an independent component).

The Connection type has an internal `closeConnectionWork`, which schedules the closing of the connection, and it runs when Reactor-thread reacts to the "Read-event raised by the dispatcher Pipe", this can also indirectly "finalize" the links (RRChannel links are one of them). But `closeConnectionWork` asynchronously waits for 2 or 3 to happen as it is an operation later in the chain

So I'm not sure about the chances real deadlock, maybe missed something, but point is, it is very important RRChannel close to happen, as we have seen this being a culprit in couple of incidents.